src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+using System.Numerics;
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal partial class CborReader
+    {
+        public CborTag ReadTag()
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+            return tag;
+        }
+
+        public void ReadTag(CborTag expectedTag)
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            if (expectedTag != tag)
+            {
+                throw new InvalidOperationException("CBOR tag mismatch.");
+            }
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+        }
+
+        public CborTag PeekTag() => PeekTagCore(out int _);
```

Comment:
I don't think I see any tests for PeekTag... presumably a good test case is to encode the same DateTimeOffset as both a string and the Unix epoch seconds and do the correct read just using the reader API.

Tests for unmapped (future) tag values are also good.  Like, write the code to read this future schema:

```
Tag value 0xFFFFFFFF_FFFFFFFF represents a date, optionally with a time of day, using a calendar other than the Gregorian calendar (a calendar date).

A calendar date is represented as a tagged array. The first value in the array is an integer identifying the calendar system from a table below. This is optionally followed by the English name of the calendar, a string with tag 119. This is optionally followed by the natural language name of the calendar, a string with tag 120. This is optionally followed by the name of an era (for era based calendars), a string with tag 729. This is followed by the day on the calendar.  If the calendar date represents a specific time on that date, there will be one additional element, the number of seconds indicated after the start of that date.

The date portion of a calendar date can be specified as either:
* A string, representing the date in yyyy-MM-dd syntax
* A number, representing the number of days since the beginning of the calendar era.

[for a challenge, the table of calendar IDs has been omitted]
```

Assuming I did all of my math right, an array of 4 such values is

84DBFFFFFFFFFFFFFFFF8405D9031066E698ADE5928C1A5017200019ECDBDBFFFFFFFFFFFFFFFF83026A353732392D31312D303919ECDBDBFFFFFFFFFFFFFFFF8406D878782DD9B1D984D8AAD98ED991D982D992D988D990D98AD98520D9B1D984D992D987D990D8ACD992D8B1D990D98AD9916A313338392D30352D313019ECDBDBFFFFFFFFFFFFFFFF8407D8776B556D6D20616C2D517572611A82395A801989C6

Reply:
Out of curiosity, where did you find that schema?